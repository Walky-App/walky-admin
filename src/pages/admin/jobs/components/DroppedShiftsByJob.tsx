import { useEffect, useState } from 'react'

import { Link } from 'react-router-dom'

import { format as formatTz, toZonedTime } from 'date-fns-tz'
import { Column } from 'primereact/column'
import { DataTable } from 'primereact/datatable'

import { HTLoadingLogo } from '../../../../components/shared/HTLoadingLogo'
import { requestService } from '../../../../services/requestServiceNew'
import { type IShift } from '../../../employee/jobs/MyJobs'

export const DroppedShiftsByJob = ({ jobId }: { jobId: string }) => {
  const [droppedShifts, setDroppedShifts] = useState<IShift[]>([])
  const [loading, setLoading] = useState(false)

  useEffect(() => {
    setLoading(true)
    const getDroppedShifts = async () => {
      try {
        const response = await requestService({ path: `shifts/by-job/${jobId}/dropped` })
        if (response.ok) {
          const data = await response.json()
          setDroppedShifts(data)
          setLoading(false)
        }
      } catch (error) {
        console.error('Failed to fetch dropped shifts:', error)
      }
    }
    getDroppedShifts()
  }, [jobId])

  return (
    <div className="card rounded-l p-6 shadow">
      {!loading ? (
        droppedShifts.length > 0 ? (
          <DataTable value={droppedShifts} className="w-full">
            <Column
              field="shift_day"
              header="Shift Day"
              body={(shift: IShift) => {
                const zonedDate = toZonedTime(shift.shift_day, shift.job_id.facility.timezone)
                return formatTz(zonedDate, 'yyyy-MM-dd, EEEE', { timeZone: shift.job_id.facility.timezone })
              }}
            />
            <Column
              header="Drop Shift Details"
              body={(shift: IShift) =>
                shift.change_history?.map((history, index) => {
                  const zonedDate = toZonedTime(history.date, shift.job_id.facility.timezone)
                  const formattedDate = formatTz(zonedDate, 'yyyy-MM-dd, EEEE', {
                    timeZone: shift.job_id.facility.timezone,
                  })

                  return (
                    <Link to={`/admin/users/employees/${history.user_id._id}`} key={index}>
                      <div className="my-4 space-y-2 border-b-2 border-t-2 border-gray-200 pb-4 pt-4 transition-colors duration-200 hover:bg-gray-300 hover:bg-opacity-50">
                        <div className="grid grid-cols-2 gap-4">
                          <p className="font-bold">
                            Full Name:&nbsp;
                            <span className="font-normal">{`${history.user_id.first_name} ${history.user_id.last_name}`}</span>
                          </p>
                          <p className="font-bold">
                            Dropped By:&nbsp;
                            <span className="font-normal">
                              {history.action === 'dropped employee' ? 'Employee' : 'Admin'}
                            </span>
                          </p>
                          <p className="font-bold">
                            When was dropped: <span className="font-normal">{formattedDate}</span>
                          </p>
                          {history.reason ? (
                            <p className="col-span-2 font-bold">
                              Reason: <span className="font-normal">{history.reason}</span>
                            </p>
                          ) : null}
                        </div>
                      </div>
                    </Link>
                  )
                })
              }
            />
          </DataTable>
        ) : (
          <h2 className="text-2xl font-bold">No Dropped Shifts!</h2>
        )
      ) : (
        <HTLoadingLogo />
      )}
    </div>
  )
}
