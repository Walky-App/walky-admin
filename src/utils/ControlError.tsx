import type { FieldErrors } from 'react-hook-form'

export function getFormErrorMessage(path: string, errors: FieldErrors) {
  const pathParts = path.split('.')
  let error: FieldErrors = errors

  for (const part of pathParts) {
    if (typeof error !== 'object' || error === null) {
      return null
    }
    error = error[part as keyof typeof error] as FieldErrors
  }

  if (error?.message) {
    return error.message ? <p className="mt-2 text-sm text-red-600">{String(error.message)}</p> : null
  }

  return null
}
