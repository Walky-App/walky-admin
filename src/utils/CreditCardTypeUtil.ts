export const getCardType = (number: string) => {
  const firstDigit = number[0]
  const firstTwoDigits = number.slice(0, 2)
  const firstFourDigits = number.slice(0, 4)

  if (/^4/.test(firstDigit)) {
    return 'Visa'
  } else if (/^5[1-5]/.test(firstTwoDigits)) {
    return 'MasterCard'
  } else if (/^3[47]/.test(firstTwoDigits)) {
    return 'Amex'
  } else if (/^6(?:011|5)/.test(firstFourDigits)) {
    return 'Discover'
  } else {
    return 'Unknown'
  }
}
