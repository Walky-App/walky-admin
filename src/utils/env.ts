export const getEnv = () => {
  return {
    VITE_API_BASE_URL: import.meta.env.VITE_API_BASE_URL || 'https://staging.walkyapp.com/api'
  };
};
