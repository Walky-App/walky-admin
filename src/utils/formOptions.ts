export type INotificationPreference = 'notification_email' | 'notification_sms'

export const notifications_preferences: INotificationPreference[] = ['notification_email', 'notification_sms']

export const job_preferences = [
  'Harvester',
  'Budtender',
  'Trimmer',
  'Packager',
  'Gardener',
  'Cultivator',
  'Extractor',
  'Front desk',
  'Greeter',
  'Id checker',
  'Inventory',
  'Data entry',
  'Event staff',
  'Promo representative',
  'Cleaning',
  'Joint roller',
  'Grow tech',
  'Clone tech',
  'Sign spinner',
]

export const facilityContactRoles = [
  { label: 'Owner', value: 'Owner' },
  { label: 'AP', value: 'AP' },
  { label: 'Onsite', value: 'Onsite' },
  { label: 'Security', value: 'Security' },
  { label: 'Manager', value: 'Manager' },
  { label: 'Other', value: 'Other' },
]

export const services = ['Trimming', 'Harvesting', 'Packaging', 'Budtending', 'Gardening', 'General Labor', 'Other']

export const countries = [
  { label: 'United States', value: 'US' },
  { label: 'Canada', value: 'CA' },
  { label: 'Mexico', value: 'MX' },
]

export const states = [
  { label: 'Alabama', value: 'AL' },
  { label: 'Alaska', value: 'AK' },
  { label: 'Arizona', value: 'AZ' },
  { label: 'Arkansas', value: 'AR' },
  { label: 'California', value: 'CA' },
  { label: 'Colorado', value: 'CO' },
  { label: 'Connecticut', value: 'CT' },
  { label: 'Delaware', value: 'DE' },
  { label: 'Florida', value: 'FL' },
  { label: 'Georgia', value: 'GA' },
  { label: 'Hawaii', value: 'HI' },
  { label: 'Idaho', value: 'ID' },
  { label: 'Illinois', value: 'IL' },
  { label: 'Indiana', value: 'IN' },
  { label: 'Iowa', value: 'IA' },
  { label: 'Kansas', value: 'KS' },
  { label: 'Kentucky', value: 'KY' },
  { label: 'Louisiana', value: 'LA' },
  { label: 'Maine', value: 'ME' },
  { label: 'Maryland', value: 'MD' },
  { label: 'Massachusetts', value: 'MA' },
  { label: 'Michigan', value: 'MI' },
  { label: 'Minnesota', value: 'MN' },
  { label: 'Mississippi', value: 'MS' },
  { label: 'Missouri', value: 'MO' },
  { label: 'Montana', value: 'MT' },
  { label: 'Nebraska', value: 'NE' },
  { label: 'Nevada', value: 'NV' },
  { label: 'New Hampshire', value: 'NH' },
  { label: 'New Jersey', value: 'NJ' },
  { label: 'New Mexico', value: 'NM' },
  { label: 'New York', value: 'NY' },
  { label: 'North Carolina', value: 'NC' },
  { label: 'North Dakota', value: 'ND' },
  { label: 'Ohio', value: 'OH' },
  { label: 'Oklahoma', value: 'OK' },
  { label: 'Oregon', value: 'OR' },
  { label: 'Pennsylvania', value: 'PA' },
  { label: 'Rhode Island', value: 'RI' },
  { label: 'South Carolina', value: 'SC' },
  { label: 'South Dakota', value: 'SD' },
  { label: 'Tennessee', value: 'TN' },
  { label: 'Texas', value: 'TX' },
  { label: 'Utah', value: 'UT' },
  { label: 'Vermont', value: 'VT' },
  { label: 'Virginia', value: 'VA' },
  { label: 'Washington', value: 'WA' },
  { label: 'West Virginia', value: 'WV' },
  { label: 'Wisconsin', value: 'WI' },
  { label: 'Wyoming', value: 'WY' },
]

export const jobTitlesOptions = [
  { title: 'Harvester', value: 'Harvester' },
  { title: 'Budtender', value: 'Budtender' },
  { title: 'Trimmer', value: 'Trimmer' },
  { title: 'Packager', value: 'Packager' },
  { title: 'Extractor', value: 'Extractor' },
  { title: 'Front desk', value: 'Front desk' },
  { title: 'Greeter', value: 'Greeter' },
  { title: 'Id checker', value: 'Id checker' },
  { title: 'Inventory', value: 'Inventory' },
  { title: 'Data entry', value: 'Data entry' },
  { title: 'Event staff', value: 'Event staff' },
  { title: 'Promo representative', value: 'Promo representative' },
  { title: 'Cleaning', value: 'Cleaning' },
  { title: 'Joint roller', value: 'Joint roller' },
  { title: 'Grow tech', value: 'Grow tech' },
  { title: 'Clone tech', value: 'Clone tech' },
  { title: 'Sign spinner', value: 'Sign spinner' },
]

export const lunchTimeOptions = [
  { label: '30 minutes', value: 30 },
  { label: '45 minutes', value: 45 },
  { label: '60 minutes', value: 60 },
]

export const jobTipsOptions = [
  { label: 'Change Required Upon Entry', value: 'Change Required Upon Entry' },
  { label: 'Lunch Included', value: 'Lunch Included' },
  { label: 'Lunch Room Available', value: 'Lunch Room Available' },
  { label: 'Lunch Will Be Off-Premise', value: 'Lunch Will Be Off-Premise' },
  { label: 'Pack a Lunch', value: 'Pack a Lunch' },
  { label: 'Parking on Street', value: 'Parking on Street' },
  { label: 'Parking Onsite', value: 'Parking Onsite' },
  { label: 'Required Identification', value: 'Required Identification' },
  { label: 'Special Equipment', value: 'Special Equipment' },
  { label: 'No gas stations nearby', value: 'No gas stations nearby' },
  { label: 'Water is provided', value: 'Water is provided' },
  { label: 'Outdoor sun exposure', value: 'Outdoor sun exposure' },
  { label: 'Must be able to lift 50 lbs', value: 'Must be able to lift 50 lbs' },
  { label: 'Steel toe shoes', value: 'Steel toe shoes' },
  { label: 'Lab coat Provided', value: 'Lab coat Provided' },
  { label: 'Head / Beard net required', value: 'Head / Beard net required' },
]
