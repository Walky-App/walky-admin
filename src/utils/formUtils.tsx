import { type FieldErrors } from 'react-hook-form'

export function getFormErrorMessage(path: string, errors: FieldErrors) {
  const pathParts = path.split('.')
  let error: FieldErrors = errors

  for (const part of pathParts) {
    if (typeof error !== 'object' || error === null) {
      return null
    }
    error = error[part as keyof typeof error] as FieldErrors
  }

  if (error?.message) {
    return <p className="mt-2 text-sm text-red-600">{String(error.message)}</p>
  }

  return null
}

export const requiredFieldsNoticeText = (
  <p className="mt-1 text-sm leading-6 text-gray-600">
    <span style={{ color: 'red' }}>*</span> indicates a required field.
  </p>
)
