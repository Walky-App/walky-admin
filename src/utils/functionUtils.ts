/* eslint-disable @typescript-eslint/no-explicit-any */
export function secondsToTimeDescription(seconds: number) {
  const minutes = Math.floor(seconds / 60)
  const hours = Math.floor(minutes / 60)

  if (hours >= 1 && minutes % 60 === 0) {
    return `${hours} hour${hours > 1 ? 's' : ''}`
  } else if (hours >= 1 && minutes % 60 !== 0) {
    const remainingMinutes = minutes % 60
    return `${hours}.${remainingMinutes} hours`
  } else {
    return `${minutes} minute${minutes !== 1 ? 's' : ''}`
  }
}

export function getModifiedProperties(oldObj: any, newObj: any): Partial<any> {
  const modifiedProperties: Partial<any> = {}
  for (const key of Object.keys(newObj)) {
    if (newObj[key] !== oldObj[key]) {
      modifiedProperties[key] = newObj[key]
    }
  }
  return modifiedProperties
}
