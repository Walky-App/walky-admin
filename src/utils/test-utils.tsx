import React, { type ReactElement } from 'react'

import { MemoryRouter } from 'react-router-dom'

import '@testing-library/jest-dom'
import { render, type RenderOptions } from '@testing-library/react'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  const originalConsoleError = console.error
  const jsDomCssError = 'Error: Could not parse CSS stylesheet'
  console.error = (...params) => {
    if (!params.find(p => p.toString().includes(jsDomCssError))) {
      originalConsoleError(...params)
    }
  }
  return <MemoryRouter>{children}</MemoryRouter>
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
