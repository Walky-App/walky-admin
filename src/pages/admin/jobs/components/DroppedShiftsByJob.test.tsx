import { render } from '../../../../utils/test-utils'
import { DroppedShiftsByJob } from './DroppedShiftsByJob'

describe('Dropped Shift By Job Component', () => {
  it('should render dropped shifts by job component', () => {
    render(<DroppedShiftsByJob jobId="qet135taetwe" />)
  })

  it('should match snapshot', () => {
    const { asFragment } = render(<DroppedShiftsByJob jobId="qet135taetwe" />)
    expect(asFragment()).toMatchSnapshot()
  })
})
