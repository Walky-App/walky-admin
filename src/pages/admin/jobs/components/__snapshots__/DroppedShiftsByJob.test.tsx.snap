// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropped Shift By Job Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="card rounded-l p-6 shadow"
  >
    <div
      class="flex items-center justify-center"
    >
      <img
        alt="Hemp Temps Loading"
        class="h-64"
        src="/assets/ht-loading.gif"
      />
    </div>
  </div>
</DocumentFragment>
`;
